2020/day4/solution.py

```diff
@@ -0,0 +1,57 @@
+import re
+kvp_matcher = re.compile(r"(\S+):(\S+)")
+
+class Passport:
+    RequiredPropertyValidatorMap = {
+        'byr' : lambda x: int(x) >= 1920 and int(x) <= 2002,
+        'iyr' : lambda x: int(x) >= 2010 and int(x) <= 2020,
+        'eyr' : lambda x: int(x) >= 2020 and int(x) <= 2030,
+        'hgt' : lambda x: (len(x) == 4 or len(x) == 5) and ((x[-2:] == "in" and int(x[:-2]) >= 59 and int(x[:-2]) <= 76) or (x[-2:] == "cm" and int(x[:-2]) >= 150 and int(x[:-2]) <= 193)),
+        'hcl' : lambda x: len(x) == 7 and x[0] == "#" and all(map(lambda y: y in "0123456789abcdef" , x[1:])),
+        'ecl' : lambda x: x in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'],
+        'pid' : lambda x: len(x) == 9 and all(map(lambda y: y in "0123456789" , x)),
+    }
+
+    OptionalProperties = [
+        'cid',
+    ]
+
+    def __init__(self):
+        self._properties = {}
+
+    def addLine(self, line):
+        kvpairs = map(lambda x: x.strip(), line.split(" "))
+        for kvp in kvpairs:
+            match = kvp_matcher.match(kvp)
+            self._properties[match.group(1)] = match.group(2)
+
+    def isValid(self):
+        for prop in Passport.RequiredPropertyValidatorMap:
+            if prop not in self._properties:
+                return False
+        return True
+
+    def isDataValid(self):
+        if not self.isValid():
+            return False
+        for prop in Passport.RequiredPropertyValidatorMap:
+            if not Passport.RequiredPropertyValidatorMap[prop](self._properties[prop]):
+                return False
+        return True
+
+file = open('input.txt')
+lines = list(map(lambda x : x.strip(), file.readlines()))
+index = 0
+validCount = 0
+dataValidCount = 0
+while index < len(lines):
+    passport = Passport()
+    while index < len(lines) and lines[index] != "":
```

Comment:
2 thoughts, both of which are slight overkill given the small amount of code here:
a) I don't see a benefit to having Passport read each line separately since it will store them all the data in memory by the end. Having the Passport constructor take the full list of strings a) makes it easier to avoid mutability problems b) avoids invalid states c) is bit nicer API. An iterative approach makes more sense when the state being kept is minimal(think of Hashers that have constant size internal state)
b) When you have a more complex loop like this, it can be helpful to split it into multiple loops. This enables separating simple for loops from more complex chunking loops and also lets you do less index match. A probably-overly-code-golfed-example:
```
import itertools

def get_passports():
    with open('input.txt') as f:
        lines = (line.strip() for line in f)
        for line in lines:
            yield Passport([line] + list(itertools.takewhile(bool, lines)))

# Can now use simple for loop/comprehensions to compute validity counts
```